xgcm/grid_ufunc.py

```diff
@@ -0,0 +1,198 @@
+        def _as_grid_ufunc_wrapper(*args, **kwargs):
+            return apply_grid_ufunc(
+                func, *args, grid=grid, signature=signature, dask=dask, **kwargs
+            )
+
+        return _as_grid_ufunc_wrapper
+
+    return _as_grid_ufunc_decorator
+
+
+def apply_grid_ufunc(func, *args, grid=None, signature="", dask="forbidden", **kwargs):
+    """
+    Apply a function to the given arguments in a grid-aware manner.
+
+    The relationship between xgcm axes on the input and output are specified by
+    `signature`. Wraps xarray.apply_ufunc, but determines the core dimensions
+    from the grid and signature passed.
+
+    Parameters
+    ----------
+    func : callable
+        Function to call like `func(*args, **kwargs)` on numpy-like unlabled
+        arrays (`.data`).
+
+        Passed directly on to `xarray.apply_ufunc`.
+    grid : xgcm.Grid
+    signature : string
+        Grid universal function signature. Specifies the xgcm.Axis names and
+        positions for each input and output variable, e.g.,
+
+        ``"(X:center)->(X:left)"`` for ``diff_center_to_left(a)`.
+    dask : {"forbidden", "allowed", "parallelized"}, default: "forbidden"
+        How to handle applying to objects containing lazy data in the form of
+        dask arrays. Passed straight on to `xarray.apply_ufunc`.
+
+    Returns
+    -------
+    result
+        The result of the call to `apply_ufunc`, but including the coordinates
+        given by the signature, read from the grid.
+    """
+
+    # Extract Axes information from signature
+    in_ax_names, out_ax_names, in_ax_pos, out_ax_pos = _parse_grid_ufunc_signature(
+        signature
+    )
+
+    # Check that input args are in correct grid positions
+    for i, (arg_ns, arg_ps, arg) in enumerate(zip(in_ax_names, in_ax_pos, args)):
+        for n, p in zip(arg_ns, arg_ps):
+            if grid.axes[n].coords[p] not in arg.coords:
+                raise ValueError(
+                    f"Mismatch between signature and input argument {i}: "
+                    f"Signature specified data to lie at Axis Position ({n}:{p}), "
+                    f"but grid coordinate {grid.axes[n].coords[p]} does not appear in argument"
+                    f"{arg}"
+                )
+
+            # TODO also check that dims are the right length for their stated Axis positions on inputs?
+
+    # Determine core dimensions for apply_ufunc
+    in_core_dims = [
+        [grid.axes[n].coords[p] for n, p in zip(arg_ns, arg_ps)]
+        for arg_ns, arg_ps in zip(in_ax_names, in_ax_pos)
+    ]
+    out_core_dims = [
+        [grid.axes[n].coords[p] for n, p in zip(arg_ns, arg_ps)]
+        for arg_ns, arg_ps in zip(out_ax_names, out_ax_pos)
+    ]
+
+    all_out_core_dims = set(dim for arg in out_core_dims for dim in arg)
+
+    # Determine expected output dimension sizes from grid._ds
+    # Only required when dask='parallelized'
+    out_sizes = {out_dim: grid._ds.dims[out_dim] for out_dim in all_out_core_dims}
```

Comment:
Is there any reason why I can't just do this to find the lengths of the output core dimensions? This should work so long as the output `(Axis:Position)` pairs are represented in the `Grid`.

Reply:
Looks reasonable to me!